Userland/Libraries/LibPDF/Function.cpp

```diff
@@ -18,11 +18,121 @@ struct Bound {
     virtual PDFErrorOr<ReadonlySpan<float>> evaluate(ReadonlySpan<float>) const override;
+
+private:
+    Vector<Bound> m_domain;
+    Vector<Bound> m_range;
+
+    Vector<unsigned> m_size;
+    int m_bits_per_sample;
+
+    enum class Order {
+        Linear = 1,
+        Cubic = 3,
+    };
+    Order m_order;
+
+    Vector<Bound> m_encode;
+    Vector<Bound> m_decode;
 };
 
+PDFErrorOr<NonnullRefPtr<SampledFunction>>
+SampledFunction::create(Document* document, Vector<Bound> domain, Optional<Vector<Bound>> range, NonnullRefPtr<StreamObject> stream)
+{
+    if (!range.has_value())
+        return Error { Error::Type::MalformedPDF, "Function type 0 requires range" };
+
+    // "TABLE 3.36 Additional entries specific to a type 0 function dictionary"
+    auto const& dict = stream->dict();
+
+    if (!dict->contains(CommonNames::Size))
+        return Error { Error::Type::MalformedPDF, "Function type 0 requires /Size" };
+    auto size_array = TRY(dict->get_array(document, CommonNames::Size));
+    Vector<unsigned> size;
+    for (auto const& size_value : *size_array) {
+        if (size_value.to_int() <= 0)
+            return Error { Error::Type::MalformedPDF, "Function type 0 /Size entry not positive" };
+        size.append(static_cast<unsigned>(size_value.to_int()));
+    }
+    if (size.size() != domain.size())
+        return Error { Error::Type::MalformedPDF, "Function type 0 /Size array has invalid size" };
+
+    if (!dict->contains(CommonNames::BitsPerSample))
+        return Error { Error::Type::MalformedPDF, "Function type 0 requires /BitsPerSample" };
+    auto bits_per_sample = TRY(document->resolve_to<int>(dict->get_value(CommonNames::BitsPerSample)));
+    switch (bits_per_sample) {
+    case 1:
+    case 2:
+    case 4:
+    case 8:
+    case 12:
+    case 16:
+    case 24:
+    case 32:
+        // Ok!
+        break;
+    default:
+        dbgln("invalid /BitsPerSample {}", bits_per_sample);
+        return Error { Error::Type::MalformedPDF, "Function type 0 has invalid /BitsPerSample" };
+    }
+
+    Order order = Order::Linear;
+    if (dict->contains(CommonNames::Order))
+        order = static_cast<Order>(TRY(document->resolve_to<int>(dict->get_value(CommonNames::Order))));
+    if (order != Order::Linear && order != Order::Cubic)
+        return Error { Error::Type::MalformedPDF, "Function type 0 has invalid /Order" };
+
+    Vector<Bound> encode;
+    if (dict->contains(CommonNames::Encode)) {
+        auto encode_array = TRY(dict->get_array(document, CommonNames::Encode));
+        if (encode_array->size() % 2 != 0)
+            return Error { Error::Type::MalformedPDF, "Function type 0 /Encode size not multiple of 2" };
+        for (size_t i = 0; i < encode_array->size(); i += 2)
+            encode.append({ encode_array->at(i).to_float(), encode_array->at(i + 1).to_float() });
+    } else {
+        for (unsigned const size : size)
```

Comment:
Both variables being called `size` is giving me the heebie-jeebies.

Reply:
👻